source/sphinx_extensions/plugin_directory/templates/action.rst

```diff
@@ -1,48 +1,21 @@
 {{ title }}
 {{ '=' * title|length }}
 
-{{ action.description }}
-
 .. raw:: html
 
-   <table class="table action-signature">
-     {% for group, specs in (('Inputs', input_specs),
-                             ('Parameters', parameter_specs),
-                             ('Outputs', output_specs)) %}
-     <thead>
-       <tr>
-         <th colspan="4">{{ group }}</th>
-       </tr>
-       {% if specs %}
-       <tr>
-         <th>Name</th>
-         <th>Type</th>
-         <th>Default</th>
-         <th>Description</th>
-       </tr>
-       {% else %}
-       <tr>
-         <th colspan="4" class="text-muted">N/A</th>
-       </tr>
-       {% endif %}
-     </thead>
-     <tbody>
-       {% for spec in specs %}
-       <tr>
-         {% for content in spec[:3] %}
-         <td>
-           <code class="docutils literal">
-             <span class="pre">{{ content }}</span>
-           </code>
-         </td>
-         {% endfor %}
-         <td>
-           {% for line in spec[3].splitlines() %}
-           {{ line|urlize }}<br/>
-           {% endfor %}
-         </td>
-       </tr>
-       {% endfor %}
-     </tbody>
-     {% endfor %}
-   </table>
+   <div class="tabbed">
+      <ul class="nav nav-tabs">
+         <li class="active"><a data-toggle="tab" href="#cli">Command line interface</a></li>
+         <li><a data-toggle="tab" href="#api">Artifact API</a></li>
+      </ul>
+      <div class="tab-content">
+         <div id="cli" class="tab-pane fade in active">
+            <pre>
+{{ cli_help }}   </pre>
```

Comment:
The trailing `</pre>` tags are gross -- I couldn't find a way to have each line in `cli_help` indented by at least 3 spaces so that it's valid RST, while also not adding trailing blank lines in the rendered `<pre>` content (which is what happens if I put `</pre>` on its own line). If anyone knows how to clean this up that would be awesome!

Reply:
How about using the jinja whitespace control flags (`-`) to strip leading and trailing whitespace, plus a judicious `strip()` for good measure:

```diff
diff --git a/source/sphinx_extensions/plugin_directory/extension.py b/source/sphinx_extensions/plugin_directory/extension.py
index 11ff4d0..1eafc46 100644
--- a/source/sphinx_extensions/plugin_directory/extension.py
+++ b/source/sphinx_extensions/plugin_directory/extension.py
@@ -67,7 +67,7 @@ def generate_rst(app):
                                       stdout=subprocess.PIPE,
                                       stderr=subprocess.STDOUT)
                 cli_help = textwrap.indent(proc.stdout.decode('utf-8'),
-                                           directive_indent)
+                                           directive_indent).strip()
                 api_help = textwrap.indent(action.__call__.__doc__,
                                            directive_indent)

diff --git a/source/sphinx_extensions/plugin_directory/templates/action.rst b/source/sphinx_extensions/plugin_directory/templates/action.rst
index 8b831c6..3e9d90c 100644
--- a/source/sphinx_extensions/plugin_directory/templates/action.rst
+++ b/source/sphinx_extensions/plugin_directory/templates/action.rst
@@ -11,7 +11,8 @@
       <div class="tab-content">
          <div id="cli" class="tab-pane fade in active">
             <pre>
-{{ cli_help }}   </pre>
+              {{- cli_help -}}
+            </pre>
          </div>
          <div id="api" class="tab-pane fade">
             <pre>
```

Now the markup looks nice:

<img width="403" alt="screen shot 2017-09-26 at 9 16 31 pm" src="https://user-images.githubusercontent.com/274668/30895574-039acbfa-a300-11e7-8374-468b50ae68c6.png">

The rendered tab looks the exact same (no extra newlines):
<img width="805" alt="screen shot 2017-09-26 at 9 26 23 pm" src="https://user-images.githubusercontent.com/274668/30895801-67bb61b6-a301-11e7-96ea-16b9c4999e31.png">

The rendered HTML:

<img width="687" alt="screen shot 2017-09-26 at 9 25 17 pm" src="https://user-images.githubusercontent.com/274668/30895788-48e77c52-a301-11e7-96a4-8876a4bcca1d.png">

I'm not sure if it is worth changing the code in this PR though.

Reply:
This is perfect, thanks for the fix @thermokarst! I pushed up those changes.